### PR TITLE
Have S.L.Expressions produce shorter IL for integral NegateChecked

### PR DESCRIPTION
**Pull request by JonHanna**
When non-nullable there's an unnecessary cast of 0 for short and sbyte, and an unnecessary store/load that can be removed.

When nullable the output is roughly like:

```
T? temp = Evaluate(operand);
T? zero = (T?)(T)0;
result = temp.HasValue & zero.HasValue
    ? (T?)(zero.GetValueOrDefault() - temp.GetValueOrDefault())
    : default(T?);
```

Instead produce code closer to:

```
T? temp = Evaluate(operand);
result = temp.GetValueOrDefault() != 0
    ? (T?)(0 - temp.GetValueOrDefault())
    : temp;

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Unary.cs [hunk 1]
```diff
@@ -72,14 +72,44 @@ private void EmitUnary(UnaryExpression node, CompilationFlags flags)
...
+
+                    // if null then push back on stack
+                    _ilg.MarkLabel(nullOrZero);
+                    _ilg.Emit(OpCodes.Ldloc, loc);
+                    FreeLocal(loc);
+                    _ilg.MarkLabel(end);
+                }
+                else
+                {
+                    _ilg.EmitDefault(type, locals: null); // locals won't be used.
```
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Unary.cs]: Can we have a simple test that validates this codegen - like in https://github.com/dotnet/corefx/commit/f1f20ecbde6e4bbe3cc982ac56558c332c2a422b

That would be useful to visualize the result and make sure there are no regressions.

Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Unary.cs]: Will do.

Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Unary.cs]: Added.

Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Unary.cs]: Thanks!!

**Comment by JonHanna** [comment]
Test UWP NETNative x86 Release Build

(This seems to be a real problem, but it's a confusing one, so giving it another shot before I look at what might be wrong).

**Comment by VSadov** [comment]
aot   tests failed since they cannot do VerifyIL.

VerifyIL needs `#if FEATURE_COMPILE` 

or, perhaps, it would make sense to make VerifyIL itself    `[Conditional(FEATURE_COMPILE)]`  ?

**Comment by JonHanna** [comment]
D'oh!

**Comment by JonHanna** [comment]
`[Conditional(FEATURE_COMPILE)]` would be neater, but I do want to add more such tests at some point in the future and there's a point after which it'd be a significant amount of time spent running tests on AOT that don't verify anything.

**Comment by JonHanna** [comment]
Test Windows x64 Debug Build
(Catastrophic failure in CI!)

**Pull request merged**